src/main/java/org/springframework/data/redis/core/ReactiveRedisTemplate.java

```diff
@@ -190,15 +176,19 @@ public ReactiveRedisConnectionFactory getConnectionFactory() {
 
 		Assert.notNull(action, "Callback object must not be null");
 
-		ReactiveRedisConnectionFactory factory = getConnectionFactory();
-		ReactiveRedisConnection conn = factory.getReactiveConnection();
+		return Flux.usingWhen(Mono.fromSupplier(() -> {
+
+			ReactiveRedisConnectionFactory factory = getConnectionFactory();
+			ReactiveRedisConnection conn = factory.getReactiveConnection();
+			ReactiveRedisConnection connToUse = preProcessConnection(conn, false);
 
-		ReactiveRedisConnection connToUse = preProcessConnection(conn, false);
+			return (exposeConnection ? connToUse : createRedisConnectionProxy(connToUse));
+		}), conn -> {
+			Publisher<T> result = action.doInRedis(conn);
 
-		ReactiveRedisConnection connToExpose = (exposeConnection ? connToUse : createRedisConnectionProxy(connToUse));
-		Publisher<T> result = action.doInRedis(connToExpose);
+			return postProcessResult(result, conn, false);
 
-		return Flux.from(postProcessResult(result, connToUse, false)).doFinally(signal -> conn.close());
+		}, ReactiveRedisConnection::closeLater, ReactiveRedisConnection::closeLater);
```

Comment:
Is it really nonblocking call?
I see that ReactiveRedisConnection::closeLater returns Mono, but inside LettuceReactiveRedisConnection it uses call marked as blocking
https://github.com/spring-projects/spring-data-redis/blob/2eb7067e8c7e859168a281145cc46ccddb42049f/src/main/java/org/springframework/data/redis/connection/lettuce/LettuceReactiveRedisConnection.java#L405-L408 
There are situations when Subscription could be canceled in non blocking thread and this code will block nonblocking thread.

Reply:
Good catch. That's fixed now.